toolbox/relnotes/hierarchy.go

```diff
@@ -0,0 +1,129 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/google/go-github/github"
+)
+
+type dictSIG map[string]dictArea
+type dictArea map[string]dictIssue
+type dictIssue map[int]dictPR
+type dictPR map[int]bool
+
+func hierarchicalNoteLayout(f *os.File, dict dictSIG, issueMap map[int]*github.Issue) {
+	for sig, areas := range dict {
+		f.WriteString(fmt.Sprintf(" - %s\n\n", strings.Title(sig)))
+		for area, issues := range areas {
+			f.WriteString(fmt.Sprintf("    - %s\n\n", strings.Title(area)))
+			for issue, prs := range issues {
+				if issue >= 0 {
+					f.WriteString(fmt.Sprintf("        - %s (#%d)\n", *issueMap[issue].Title, issue))
+				} else {
+					f.WriteString(fmt.Sprintf("        - NullIssue\n"))
+				}
+				for pr := range prs {
+					f.WriteString(fmt.Sprintf("            * %s (#%d, @%s)\n", extractReleaseNote(issueMap[pr]), pr, *issueMap[pr].User.Login))
+				}
+				f.WriteString("\n")
+			}
+		}
+	}
+}
+
+// createHierarchicalNote given release PRs and issue map, creates hierarchical release note
+// map[SIG]map[Area]map[Issue]PR.
+func createHierarchicalNote(prs []int, issueMap map[int]*github.Issue) dictSIG {
+	var dict = dictSIG{}
+
+	for _, pr := range prs {
+		issues := extractFixedIssues(*issueMap[pr].Body)
+		if len(issues) == 0 {
+			// In our design doc, the automation should enforce every release-note PR
+			// with at least one issue. However it's observed that the rule is not
+			// applied yet. Also old PRs may not link to an issue.
+			//
+			// To produce info-richer release note, we try to get SIG and Area label
```

Comment:
Unfortunately, most of PRs don't have either sig or area label. I once felt a bit painful when I wanted to filter PRs by sig. I was thinking we enforce sig labels on PRs as we already do on issues. Then we may get more info here. @enisoc @Bradamant3 WDYT?

Reply:
Enforcing sig label on PR will also be helpful for #348.